app/src/screen.c

```diff
@@ -134,6 +140,50 @@ create_texture(SDL_Renderer *renderer, struct size frame_size) {
                              frame_size.width, frame_size.height);
 }
 
+// This may be called more than once to work around SDL bugs
+bool
+screen_init_renderer_and_texture(struct screen *screen) {
+    if (screen->texture != NULL) {
+      SDL_DestroyTexture(screen->texture);
+      screen->texture = NULL;
+    }
+    if (screen->renderer != NULL) {
+      SDL_DestroyRenderer(screen->renderer);
+      screen->renderer = NULL;
+    }
+
+    screen->renderer = SDL_CreateRenderer(screen->window, -1,
+                                            SDL_RENDERER_ACCELERATED);
+    if (!screen->renderer) {
+        LOGC("Could not create renderer: %s", SDL_GetError());
+        screen_destroy(screen);
+        return false;
+    }
+
+    if (SDL_RenderSetLogicalSize(screen->renderer, screen->frame_size.width,
+                                 screen->frame_size.height)) {
+        LOGE("Could not set renderer logical size: %s", SDL_GetError());
+        screen_destroy(screen);
+        return false;
+    }
+
+#ifdef HIDPI_SUPPORT
+    int window_w, window_h, renderer_w, renderer_h;
+    SDL_GetWindowSize(screen->window, &window_w, &window_h);
+    SDL_GetRendererOutputSize(screen->renderer, &renderer_w, &renderer_h);
+    screen->expected_hidpi_w_factor = renderer_w * 1000 / window_w;
+    screen->expected_hidpi_h_factor = renderer_h * 1000 / window_h;
```

Comment:
To avoid any rounding issues, we could store two "rationals" (a struct of two `int`s) `scale_x = {renderer_w, window_w}` and `scale_y = {renderer_h, window_h}` (in pseudo-code).

Then, when we need to compare:
```
bool scale_x_changed = new_scale_x.num * old_scale_x.den == old_scale_x.num * new_scale_x.den;
bool scale_y_changed = new_scale_y.num * old_scale_y.den == old_scale_y.num * new_scale_y.den;
```

Reply:
👍